File: tools/alr-code-map/references.py
"""Reference binding from complete parsed citations; no fuzzy source matching.

Native note IDs and displayed labels are separate inputs. This model never
inserts a Word field. A pointer-preserving command returns the exact target ID;
a correction command requires a unique parsed source/alias association.
"""
from __future__ import annotations
from dataclasses import dataclass
from hashlib import sha256
from typing import Iterable
from citations import Citation, LegalLocator


@dataclass(frozen=True)
class Note:
    id: str
    label: str
    scope: str
    citations: tuple[Citation, ...]


@dataclass(frozen=True)
class Binding:
    note_id: str
    occurrence: int
    source: tuple[str, ...]
    basis: str | None
    locator: tuple | LegalLocator | None
    qualifications: tuple[str, ...]
    origin_note: str

    @property
    def meaning(self):
        return self.source, self.basis, self.locator, self.qualifications


@dataclass(frozen=True)
class Graph:
    fingerprint: str
    notes: tuple[Note, ...]
    bindings: dict[tuple[str, int], Binding]
    unresolved: dict[tuple[str, int], str]
    aliases: dict[str, tuple[tuple[str, ...], ...]]

    def preserve_pointer(self, scope: str, label: str) -> str:
        targets = [n.id for n in self.notes if n.scope == scope and n.label == label]
        if len(targets) != 1: raise ValueError('display label has no unique native target')
        return targets[0]

    def correct_alias_target(self, note_id: str, alias: str) -> str:
        """Find first preceding *full* citation of one exact declared source.

        Corrects a source association only, not legal relevance. Names/editions
        not connected by a parsed identity remain different. No same-surname rule.
        """
        sources = self.aliases.get(alias, ())
        if len(sources) != 1: raise ValueError('unknown/colliding alias')
        index = next(i for i, n in enumerate(self.notes) if n.id == note_id)
        for n in self.notes[:index]:
            if any(c.source == sources[0] for c in n.citations): return n.id
        raise ValueError('no preceding full-source declaration')


def compile_graph(notes: Iterable[Note]) -> Graph:
    notes = tuple(notes)
    if not notes or any(not n.id or not n.scope or not n.label for n in notes):
        raise ValueError('complete native note identities and numbering scopes required')
    if len({n.id for n in notes}) != len(notes): raise ValueError('duplicate native note ID')
    aliases = {}
    for n in notes:
        for c in n.citations:
            if c.source is not None and c.kind not in {'case','reporter','journal','book','web','legislation'}:
                raise ValueError('source key attached to an unsupported citation kind')
            if c.source is not None and (len(c.source) < 2 or any(not isinstance(s, str) for s in c.source)):
                raise ValueError('malformed parsed source key')
            if c.source is not None and c.declared_alias:
                aliases.setdefault(c.declared_alias, set()).add(c.source)
    aliases = {k: tuple(sorted(v)) for k, v in aliases.items()}
    bindings = {}; unresolved = {}
    g = Graph(sha256(repr(notes).encode()).hexdigest(), notes, bindings, unresolved, aliases)
    index = {n.id: i for i, n in enumerate(notes)}
    for i, n in enumerate(notes):
        for j, c in enumerate(n.citations):
            key = (n.id, j)
            try:
                if c.source is not None:
                    if not c.source[0] or not any(c.source[1:]): raise ValueError('empty source identity')
                    bindings[key] = Binding(n.id, j, c.source, c.basis, c.locator, c.qualifications, n.id)
                    continue
                if c.kind == 'supra':
                    target = g.preserve_pointer(n.scope, c.note_label or '')
                    if index[target] >= i: raise ValueError('supra does not target a preceding note')
                    candidates = [b for (nid, _), b in bindings.items() if nid == target]
                    # An unresolved occurrence in the target makes it incomplete.
                    if len(candidates) != len(notes[index[target]].citations):
                        raise ValueError('target note has unresolved source components')
                    if c.alias:
                        source_set = aliases.get(c.alias, ())
                        if len(source_set) != 1: raise ValueError('unknown/colliding alias')
                        candidates = [b for b in candidates if b.source == source_set[0]]
                    unique = {b.source for b in candidates}
                    if len(unique) != 1: raise ValueError('supra target source is ambiguous')
                    parent = candidates[0]
                    # Supra does not silently inherit the old pinpoint.
                    bindings[key] = Binding(n.id, j, parent.source, c.basis, c.locator,
                                            c.qualifications, parent.origin_note)
                elif c.kind == 'ibid':
                    # Restrict to first citation in current note and exactly one
                    # fully resolved citation in the immediately preceding note.
                    if j != 0 or i == 0 or len(notes[i-1].citations) != 1:
                        raise ValueError('ibid governing citation not uniquely established')
                    parent = bindings.get((notes[i-1].id, 0))
                    if parent is None: raise ValueError('ibid parent unresolved')
                    basis = c.basis if c.locator is not None else parent.basis
                    locator = c.locator if c.locator is not None else parent.locator
                    bindings[key] = Binding(n.id, j, parent.source, basis, locator,
                                            c.qualifications, parent.origin_note)
                else: raise ValueError('unsupported unresolved root')
            except ValueError as e:
                unresolved[key] = str(e)
    return g


def _locator(b: Binding) -> str:
    if b.locator is None: return ''
    if b.basis in ('page', 'paragraph'):
        vals = []
        for item in b.locator:
            if not isinstance(item, tuple) or len(item) != 2 or not all(type(x) is int and x > 0 for x in item):
                raise ValueError('unsupported numeric binding')
            a, z = item
            if z < a: raise ValueError('descending binding')
            vals.append(str(a) if a == z else f'{a}–{z}')
        label = '' if b.basis == 'page' else ('paras ' if len(vals) > 1 or any(a != z for a, z in b.locator) else 'para ')
        return ' at ' + label + ', '.join(vals)
    if b.basis == 'section' and isinstance(b.locator, LegalLocator):
        loc = b.locator
        if len(loc.separators) != len(loc.identifiers) - 1: raise ValueError('incomplete legal locator')
        label = {'section': ('s', 'ss'), 'article': ('art', 'arts'), 'rule': ('rule', 'rules')}[loc.unit][len(loc.identifiers) > 1]
        value = loc.identifiers[0]
        for sep, ident in zip(loc.separators, loc.identifiers[1:]): value += sep + ident
        return ', ' + label + ' ' + value
    raise ValueError('locator needs a lossless renderer')


def repair_ibid_after_reorder(graph: Graph, current: Iterable[Note], note_id: str, *,
                              current_graph_fingerprint: str) -> str | None:
    """Use the old binding, never reinterpret Ibid from its new neighbour.

    A caller provides a validated new native-note ordering/labels, and a fresh
    fingerprint of the old analyzed graph. Added unrelated notes are permitted;
    modifications to any retained citation invalidate this routine. None means
    the existing Ibid still has the exact tested meaning. Text result is a
    proposal, not a native mutation or an automatic acceptance of a note move.
    """
    if current_graph_fingerprint != graph.fingerprint: raise ValueError('stale reference graph')
    current = tuple(current)
    new = compile_graph(current)
    old_by = {n.id: n for n in graph.notes}
    for n in current:
        if n.id in old_by and n.citations != old_by[n.id].citations:
            raise ValueError('a retained citation changed; rebuild associations')
    old = graph.bindings.get((note_id, 0))
    if old is None: raise ValueError('old intended source is unresolved')
    original = old_by[note_id]
    if len(original.citations) != 1 or original.citations[0].kind != 'ibid':
        raise ValueError('not a single bound ibid note')
    index = next(i for i, n in enumerate(current) if n.id == note_id)
    rebound = new.bindings.get((note_id, 0))
    if rebound and rebound.meaning == old.meaning: return None
    origins = [n for n in current[:index] if any(c.source == old.source for c in n.citations)]
    if not origins: raise ValueError('original source has no preceding full citation')
    target = origins[0]
    if not target.label.isascii() or not target.label.isdecimal(): raise ValueError('supra label is not ordinary Arabic numbering')
    if new.preserve_pointer(current[index].scope, target.label) != target.id:
        raise ValueError('cross-scope numbering is ambiguous')
    aliases = sorted(k for k, v in graph.aliases.items() if v == (old.source,))
    if not aliases: raise ValueError('no unique declared short form for old source')
    suffix = ' [' + ', '.join(old.qualifications) + ']' if old.qualifications else ''
    return aliases[0] + ', supra note ' + target.label + _locator(old) + suffix + '.'

File: internal/alrmap/references.go
package alrmap

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"sort"
	"strconv"
	"strings"

	"github.com/eliziff/WordUp/internal/alrmap/citations"
	"github.com/eliziff/WordUp/internal/alrmap/internal/pytext"
)

// Note is one native note: its persistent occurrence identity (ID), its
// currently displayed number (Label), its numbering scope and its parsed
// citations. Native note IDs and displayed labels are separate inputs.
type Note struct {
	ID        string
	Label     string
	Scope     string
	Citations []citations.Citation
}

// Binding is one citation occurrence resolved to a source.
type Binding struct {
	NoteID         string
	Occurrence     int
	Source         []string
	Basis          string
	Locator        *citations.Locator
	Qualifications []string
	OriginNote     string
}

// Meaning is the comparable (source, basis, locator, qualifications) tuple.
func (b Binding) Meaning() string {
	return strings.Join([]string{citations.SourceKey(b.Source), b.Basis, b.Locator.String(), strings.Join(b.Qualifications, "\x1f")}, "\x1e")
}

// BindingKey addresses one citation occurrence inside a note.
type BindingKey struct {
	NoteID     string
	Occurrence int
}

// Graph is the compiled reference graph. This model never inserts a Word
// field: a pointer-preserving command returns the exact target ID, and a
// correction command requires a unique parsed source/alias association.
type Graph struct {
	Fingerprint string
	Notes       []Note
	Bindings    map[BindingKey]Binding
	Unresolved  map[BindingKey]string
	Aliases     map[string][][]string
}

// PreservePointer returns the one native note ID displaying label in scope.
func (g *Graph) PreservePointer(scope, label string) (string, error) {
	var targets []string
	for _, n := range g.Notes {
		if n.Scope == scope && n.Label == label {
			targets = append(targets, n.ID)
		}
	}
	if len(targets) != 1 {
		return "", errors.New("display label has no unique native target")
	}
	return targets[0], nil
}

// CorrectAliasTarget finds the first preceding full citation of one exact
// declared source. It corrects a source association only, not legal
// relevance; names/editions not connected by a parsed identity remain
// different, and there is no same-surname rule.
func (g *Graph) CorrectAliasTarget(noteID, alias string) (string, error) {
	sources := g.Aliases[alias]
	if len(sources) != 1 {
		return "", errors.New("unknown/colliding alias")
	}
	index := -1
	for i, n := range g.Notes {
		if n.ID == noteID {
			index = i
			break
		}
	}
	if index < 0 {
		return "", errors.New("unknown note ID")
	}
	want := citations.SourceKey(sources[0])
	for _, n := range g.Notes[:index] {
		for _, c := range n.Citations {
			if c.Source != nil && citations.SourceKey(c.Source) == want {
				return n.ID, nil
			}
		}
	}
	return "", errors.New("no preceding full-source declaration")
}

var sourceKinds = map[string]bool{"case": true, "reporter": true, "journal": true, "book": true, "web": true, "legislation": true}

func compareSource(a, b []string) int {
	for i := 0; i < len(a) && i < len(b); i++ {
		if c := strings.Compare(a[i], b[i]); c != 0 {
			return c
		}
	}
	return len(a) - len(b)
}

// notesFingerprint is the stable identity of the analyzed note sequence
// (the reference hashes repr(notes)); it covers IDs, labels, scopes and the
// complete parsed citations including their tokens.
func notesFingerprint(notes []Note) string {
	var b strings.Builder
	for _, n := range notes {
		fmt.Fprintf(&b, "%q\x1d%q\x1d%q\x1d", n.ID, n.Label, n.Scope)
		for _, c := range n.Citations {
			fmt.Fprintf(&b, "%q\x1c", c.Meaning())
			for _, t := range c.Tokens {
				fmt.Fprintf(&b, "%q:%d:%d:%q;", t.Kind, t.Start, t.End, t.Value)
			}
			b.WriteString("\x1c")
		}
		b.WriteString("\x1d\n")
	}
	sum := sha256.Sum256([]byte(b.String()))
	return hex.EncodeToString(sum[:])
}

// CompileGraph binds complete parsed sources, exact declarations, supported
// supra pointers and supported Ibid chains. Unresolvable occurrences are
// recorded with their reason rather than guessed.
func CompileGraph(notes []Note) (*Graph, error) {
	notes = append([]Note(nil), notes...)
	if len(notes) == 0 {
		return nil, errors.New("complete native note identities and numbering scopes required")
	}
	ids := map[string]bool{}
	for _, n := range notes {
		if n.ID == "" || n.Scope == "" || n.Label == "" {
			return nil, errors.New("complete native note identities and numbering scopes required")
		}
		ids[n.ID] = true
	}
	if len(ids) != len(notes) {
		return nil, errors.New("duplicate native note ID")
	}
	aliasSets := map[string]map[string][]string{}
	for _, n := range notes {
		for _, c := range n.Citations {
			if c.Source == nil {
				continue
			}
			if !sourceKinds[c.Kind] {
				return nil, errors.New("source key attached to an unsupported citation kind")
			}
			if len(c.Source) < 2 {
				return nil, errors.New("malformed parsed source key")
			}
			if c.DeclaredAlias != "" {
				if aliasSets[c.DeclaredAlias] == nil {
					aliasSets[c.DeclaredAlias] = map[string][]string{}
				}
				aliasSets[c.DeclaredAlias][citations.SourceKey(c.Source)] = c.Source
			}
		}
	}
	aliases := map[string][][]string{}
	for k, set := range aliasSets {
		var list [][]string
		for _, s := range set {
			list = append(list, s)
		}
		sort.Slice(list, func(i, j int) bool { return compareSource(list[i], list[j]) < 0 })
		aliases[k] = list
	}
	g := &Graph{Fingerprint: notesFingerprint(notes), Notes: notes, Bindings: map[BindingKey]Binding{}, Unresolved: map[BindingKey]string{}, Aliases: aliases}
	index := map[string]int{}
	for i, n := range notes {
		index[n.ID] = i
	}
	for i, n := range notes {
		for j, c := range n.Citations {
			key := BindingKey{n.ID, j}
			if err := g.bind(notes, index, i, j, c, key); err != nil {
				g.Unresolved[key] = err.Error()
			}
		}
	}
	return g, nil
}

func (g *Graph) bind(notes []Note, index map[string]int, i, j int, c citations.Citation, key BindingKey) error {
	n := notes[i]
	if c.Source != nil {
		identified := false
		for _, s := range c.Source[1:] {
			if s != "" {
				identified = true
				break
			}
		}
		if c.Source[0] == "" || !identified {
			return errors.New("empty source identity")
		}
		g.Bindings[key] = Binding{n.ID, j, c.Source, c.Basis, c.Locator, c.Qualifications, n.ID}
		return nil
	}
	switch c.Kind {
	case "supra":
		target, err := g.PreservePointer(n.Scope, c.NoteLabel)
		if err != nil {
			return err
		}
		if index[target] >= i {
			return errors.New("supra does not target a preceding note")
		}
		targetNote := notes[index[target]]
		var candidates []Binding
		for k := range targetNote.Citations {
			if b, ok := g.Bindings[BindingKey{target, k}]; ok {
				candidates = append(candidates, b)
			}
		}
		// An unresolved occurrence in the target makes it incomplete.
		if len(candidates) != len(targetNote.Citations) {
			return errors.New("target note has unresolved source components")
		}
		if c.Alias != "" {
			sourceSet := g.Aliases[c.Alias]
			if len(sourceSet) != 1 {
				return errors.New("unknown/colliding alias")
			}
			want := citations.SourceKey(sourceSet[0])
			var filtered []Binding
			for _, b := range candidates {
				if citations.SourceKey(b.Source) == want {
					filtered = append(filtered, b)
				}
			}
			candidates = filtered
		}
		unique := map[string]bool{}
		for _, b := range candidates {
			unique[citations.SourceKey(b.Source)] = true
		}
		if len(unique) != 1 {
			return errors.New("supra target source is ambiguous")
		}
		parent := candidates[0]
		// Supra does not silently inherit the old pinpoint.
		g.Bindings[key] = Binding{n.ID, j, parent.Source, c.Basis, c.Locator, c.Qualifications, parent.OriginNote}
	case "ibid":
		// Restrict to first citation in current note and exactly one fully
		// resolved citation in the immediately preceding note.
		if j != 0 || i == 0 || len(notes[i-1].Citations) != 1 {
			return errors.New("ibid governing citation not uniquely established")
		}
		parent, ok := g.Bindings[BindingKey{notes[i-1].ID, 0}]
		if !ok {
			return errors.New("ibid parent unresolved")
		}
		basis, locator := parent.Basis, parent.Locator
		if c.Locator != nil {
			basis, locator = c.Basis, c.Locator
		}
		g.Bindings[key] = Binding{n.ID, j, parent.Source, basis, locator, c.Qualifications, parent.OriginNote}
	default:
		return errors.New("unsupported unresolved root")
	}
	return nil
}

var legalLabels = map[string][2]string{"section": {"s", "ss"}, "article": {"art", "arts"}, "rule": {"rule", "rules"}}

// renderLocator serializes a binding's locator losslessly.
func renderLocator(b Binding) (string, error) {
	if b.Locator == nil {
		return "", nil
	}
	if b.Basis == "page" || b.Basis == "paragraph" {
		if b.Locator.Legal != nil {
			return "", errors.New("unsupported numeric binding")
		}
		var vals []string
		multiple := len(b.Locator.Ranges) > 1
		for _, r := range b.Locator.Ranges {
			if r.Lo <= 0 || r.Hi <= 0 {
				return "", errors.New("unsupported numeric binding")
			}
			if r.Hi < r.Lo {
				return "", errors.New("descending binding")
			}
			if r.Lo == r.Hi {
				vals = append(vals, strconv.Itoa(r.Lo))
			} else {
				multiple = true
				vals = append(vals, strconv.Itoa(r.Lo)+"–"+strconv.Itoa(r.Hi))
			}
		}
		label := ""
		if b.Basis != "page" {
			label = "para "
			if multiple {
				label = "paras "
			}
		}
		return " at " + label + strings.Join(vals, ", "), nil
	}
	if b.Basis == "section" && b.Locator.Legal != nil {
		loc := b.Locator.Legal
		if len(loc.Separators) != len(loc.Identifiers)-1 {
			return "", errors.New("incomplete legal locator")
		}
		labels, ok := legalLabels[loc.Unit]
		if !ok {
			return "", errors.New("locator needs a lossless renderer")
		}
		label := labels[0]
		if len(loc.Identifiers) > 1 {
			label = labels[1]
		}
		value := loc.Identifiers[0]
		for i, sep := range loc.Separators {
			value += sep + loc.Identifiers[i+1]
		}
		return ", " + label + " " + value, nil
	}
	return "", errors.New("locator needs a lossless renderer")
}

func citationsEqual(a, b []citations.Citation) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if !a[i].Equal(b[i]) {
			return false
		}
	}
	return true
}

// RepairIbidAfterReorder uses the old binding, never reinterpreting Ibid from
// its new neighbour. The caller supplies the validated new native-note
// ordering/labels and a fresh fingerprint of the old analyzed graph. Added
// unrelated notes are permitted; modifications to any retained citation
// invalidate the routine. ok=false means the existing Ibid still has the
// exact tested meaning. The text is a proposal, not a native mutation.
func RepairIbidAfterReorder(graph *Graph, current []Note, noteID, currentGraphFingerprint string) (string, bool, error) {
	if currentGraphFingerprint != graph.Fingerprint {
		return "", false, errors.New("stale reference graph")
	}
	newGraph, err := CompileGraph(current)
	if err != nil {
		return "", false, err
	}
	oldBy := map[string]Note{}
	for _, n := range graph.Notes {
		oldBy[n.ID] = n
	}
	for _, n := range current {
		if old, ok := oldBy[n.ID]; ok && !citationsEqual(n.Citations, old.Citations) {
			return "", false, errors.New("a retained citation changed; rebuild associations")
		}
	}
	old, ok := graph.Bindings[BindingKey{noteID, 0}]
	if !ok {
		return "", false, errors.New("old intended source is unresolved")
	}
	original := oldBy[noteID]
	if len(original.Citations) != 1 || original.Citations[0].Kind != "ibid" {
		return "", false, errors.New("not a single bound ibid note")
	}
	index := -1
	for i, n := range current {
		if n.ID == noteID {
			index = i
			break
		}
	}
	if index < 0 {
		return "", false, errors.New("note is absent from the current ordering")
	}
	if rebound, ok := newGraph.Bindings[BindingKey{noteID, 0}]; ok && rebound.Meaning() == old.Meaning() {
		return "", false, nil
	}
	oldKey := citations.SourceKey(old.Source)
	var target *Note
	for i := range current[:index] {
		n := current[i]
		for _, c := range n.Citations {
			if c.Source != nil && citations.SourceKey(c.Source) == oldKey {
				target = &n
				break
			}
		}
		if target != nil {
			break
		}
	}
	if target == nil {
		return "", false, errors.New("original source has no preceding full citation")
	}
	if !pytext.IsDecimalDigits(target.Label) {
		return "", false, errors.New("supra label is not ordinary Arabic numbering")
	}
	pointer, err := newGraph.PreservePointer(current[index].Scope, target.Label)
	if err != nil {
		return "", false, err
	}
	if pointer != target.ID {
		return "", false, errors.New("cross-scope numbering is ambiguous")
	}
	var aliases []string
	for k, v := range graph.Aliases {
		if len(v) == 1 && citations.SourceKey(v[0]) == oldKey {
			aliases = append(aliases, k)
		}
	}
	if len(aliases) == 0 {
		return "", false, errors.New("no unique declared short form for old source")
	}
	sort.Strings(aliases)
	locator, err := renderLocator(old)
	if err != nil {
		return "", false, err
	}
	suffix := ""
	if len(old.Qualifications) > 0 {
		suffix = " [" + strings.Join(old.Qualifications, ", ") + "]"
	}
	return aliases[0] + ", supra note " + target.Label + locator + suffix + ".", true, nil
}
